File: frontend/src/services/categoryService.ts
/**
 * Category service for MyBudget.
 *
 * Handles category and category group CRUD operations.
 */

import { api } from './api'
import type {
  Category,
  CategoryGroup,
  CategoryCreate,
  CategoryUpdate,
  CategoryGroupCreate,
  CategoryGroupUpdate,
  CategoryListResponse,
} from '@/types/category'
import type { Assignment, AssignmentCreate } from '@/types/budget'

/**
 * Category service with methods for managing categories and groups.
 */
export const categoryService = {
  // Category Group operations

  /**
   * Create a new category group.
   *
   * @param data - Category group creation data
   * @returns The created category group
   */
  async createGroup(data: CategoryGroupCreate): Promise<CategoryGroup> {
    return api.post<CategoryGroup>('/categories/groups', data)
  },

  /**
   * Get all category groups for the current user.
   *
   * @returns List of category groups
   */
  async listGroups(): Promise<CategoryGroup[]> {
    return api.get<CategoryGroup[]>('/categories/groups')
  },

  /**
   * Update a category group.
   *
   * @param id - Category group ID
   * @param data - Update data
   * @returns The updated category group
   */
  async updateGroup(id: string, data: CategoryGroupUpdate): Promise<CategoryGroup> {
    return api.put<CategoryGroup>(`/categories/groups/${id}`, data)
  },

  /**
   * Delete a category group.
   *
   * @param id - Category group ID
   */
  async deleteGroup(id: string): Promise<void> {
    return api.delete(`/categories/groups/${id}`)
  },

  // Category operations

  /**
   * Create a new category.
   *
   * @param data - Category creation data
   * @returns The created category
   */
  async create(data: CategoryCreate): Promise<Category> {
    return api.post<Category>('/categories/', data)
  },

  /**
   * Get all categories with their groups.
   *
   * @returns Categories organized by groups
   */
  async list(): Promise<CategoryListResponse> {
    return api.get<CategoryListResponse>('/categories/')
  },

  /**
   * Get a single category by ID.
   *
   * @param id - Category ID
   * @returns The category
   */
  async get(id: string): Promise<Category> {
    return api.get<Category>(`/categories/${id}`)
  },

  /**
   * Update a category.
   *
   * @param id - Category ID
   * @param data - Update data
   * @returns The updated category
   */
  async update(id: string, data: CategoryUpdate): Promise<Category> {
    return api.put<Category>(`/categories/${id}`, data)
  },

  /**
   * Delete a category.
   *
   * @param id - Category ID
   */
  async delete(id: string): Promise<void> {
    return api.delete(`/categories/${id}`)
  },

  /**
   * Assign (or unassign) funds to a category.
   *
   * @param categoryId - Category ID to assign funds to
   * @param data - Assignment data (amount and month)
   * @returns The created assignment
   */
  async assignFunds(categoryId: string, data: AssignmentCreate): Promise<Assignment> {
    return api.post<Assignment>(`/categories/${categoryId}/assign`, data)
  },
}

export default categoryService

File: backend/src/mybudget/services/category_service.py
"""
Category service for business logic operations.
"""
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from mybudget.models.category import Category, CategoryGroup
from mybudget.schemas.category import (
    CategoryGroupCreate,
    CategoryGroupUpdate,
    CategoryCreate,
    CategoryUpdate,
)


class CategoryService:
    """Service for category and category group operations."""

    def __init__(self, db: AsyncSession):
        """Initialize with database session."""
        self.db = db

    # Category Group operations

    async def create_category_group(
        self, user_id: UUID, data: CategoryGroupCreate
    ) -> CategoryGroup:
        """Create a new category group."""
        group = CategoryGroup(
            user_id=user_id,
            name=data.name,
            display_order=data.display_order,
        )

        self.db.add(group)
        await self.db.commit()
        await self.db.refresh(group)

        return group

    async def get_category_group(
        self, user_id: UUID, group_id: UUID
    ) -> CategoryGroup | None:
        """Get category group by ID."""
        stmt = select(CategoryGroup).where(
            CategoryGroup.id == group_id,
            CategoryGroup.user_id == user_id,
        )
        result = await self.db.execute(stmt)
        return result.scalar_one_or_none()

    async def list_category_groups(
        self, user_id: UUID
    ) -> list[CategoryGroup]:
        """List all category groups for a user, ordered by display_order."""
        stmt = (
            select(CategoryGroup)
            .where(CategoryGroup.user_id == user_id)
            .order_by(CategoryGroup.display_order, CategoryGroup.name)
        )
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def update_category_group(
        self, user_id: UUID, group_id: UUID, data: CategoryGroupUpdate
    ) -> CategoryGroup | None:
        """Update a category group."""
        group = await self.get_category_group(user_id, group_id)
        if not group:
            return None

        if data.name is not None:
            group.name = data.name
        if data.display_order is not None:
            group.display_order = data.display_order

        await self.db.commit()
        await self.db.refresh(group)

        return group

    async def delete_category_group(
        self, user_id: UUID, group_id: UUID
    ) -> bool:
        """Delete a category group (cascades to categories)."""
        group = await self.get_category_group(user_id, group_id)
        if not group:
            return False

        await self.db.delete(group)
        await self.db.commit()

        return True

    # Category operations

    async def create_category(
        self, user_id: UUID, data: CategoryCreate
    ) -> Category | None:
        """
        Create a new category.

        Returns None if the specified group doesn't exist or belong to user.
        """
        # Verify group exists and belongs to user
        group = await self.get_category_group(user_id, data.group_id)
        if not group:
            return None

        category = Category(
            user_id=user_id,
            group_id=data.group_id,
            name=data.name,
        )

        self.db.add(category)
        await self.db.commit()
        await self.db.refresh(category)

        return category

    async def get_category(
        self, user_id: UUID, category_id: UUID
    ) -> Category | None:
        """Get category by ID."""
        stmt = select(Category).where(
            Category.id == category_id,
            Category.user_id == user_id,
        )
        result = await self.db.execute(stmt)
        return result.scalar_one_or_none()

    async def list_categories(
        self, user_id: UUID, group_id: UUID | None = None
    ) -> list[Category]:
        """List all categories for a user, optionally filtered by group."""
        stmt = select(Category).where(Category.user_id == user_id)

        if group_id is not None:
            stmt = stmt.where(Category.group_id == group_id)

        stmt = stmt.order_by(Category.name)
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def update_category(
        self, user_id: UUID, category_id: UUID, data: CategoryUpdate
    ) -> Category | None:
        """Update a category."""
        category = await self.get_category(user_id, category_id)
        if not category:
            return None

        if data.group_id is not None:
            # Verify new group exists and belongs to user
            group = await self.get_category_group(user_id, data.group_id)
            if not group:
                return None
            category.group_id = data.group_id

        if data.name is not None:
            category.name = data.name

        await self.db.commit()
        await self.db.refresh(category)

        return category

    async def delete_category(
        self, user_id: UUID, category_id: UUID
    ) -> bool:
        """Delete a category."""
        category = await self.get_category(user_id, category_id)
        if not category:
            return False

        await self.db.delete(category)
        await self.db.commit()

        return True

    async def list_groups_with_categories(
        self, user_id: UUID
    ) -> list[tuple[CategoryGroup, list[Category]]]:
        """List all category groups with their categories."""
        groups = await self.list_category_groups(user_id)
        result = []

        for group in groups:
            categories = await self.list_categories(user_id, group.id)
            result.append((group, categories))

        return result
